sdk/tests/utilities/test_api_client_factory_auth_flow.py

```diff
@@ -0,0 +1,119 @@
+import unittest
+import os
+from lusid import InstrumentsApi, ResourceListOfInstrumentIdTypeDescriptor
+from lusid.utilities import ApiClientFactory
+from utilities import CredentialsSource
+from unittest.mock import patch
+
+source_config_details, config_keys = CredentialsSource.fetch_credentials(), CredentialsSource.fetch_config_keys()
+
+
+class ApiFactory(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        # create a configured API client
+        cls.secrets = CredentialsSource.secrets_path()
+        cls.os_environ_dict_str = 'os.environ'
+
+    def validate_api(self, api):
+        result = api.get_instrument_identifier_types()
+        self.assertIsNotNone(result)
+        self.assertIsInstance(result, ResourceListOfInstrumentIdTypeDescriptor)
+        self.assertGreater(len(result.values), 0)
+
+    def get_pat_env_var(self):
+
+        pat_env_vars = {
+            "FBN_LUSID_API_URL": os.getenv("FBN_LUSID_API_URL"),
+            "FBN_LUSID_ACCESS_TOKEN": os.getenv("FBN_LUSID_ACCESS_TOKEN")
+        }
+
+        return pat_env_vars
+
+    def get_secrets_env_vars(self):
+        pat_env_vars = {
+            "FBN_LUSID_API_URL": os.getenv("FBN_LUSID_API_URL"),
+            "FBN_TOKEN_URL": os.getenv("FBN_TOKEN_URL"),
+            "FBN_USERNAME": os.getenv("FBN_USERNAME"),
+            "FBN_PASSWORD": os.getenv("FBN_PASSWORD"),
+            "FBN_CLIENT_ID": os.getenv("FBN_CLIENT_ID"),
+            "FBN_CLIENT_SECRET": os.getenv("FBN_CLIENT_SECRET")
+        }
+
+        return pat_env_vars
+
+    def test_bad_pat_in_param_but_good_pat_in_env_vars(self):
+
+        with patch.dict(self.os_environ_dict_str, self.get_pat_env_var(), clear=True):
+            factory = ApiClientFactory(token="INVALID_TOKEN")
+            token = factory.api_client.configuration.access_token
+            self.assertEqual(token, "INVALID_TOKEN")
+
+    def test_bad_secrets_file_in_param_but_good_pat_in_env_vars(self):
+
+        with patch.dict(self.os_environ_dict_str, self.get_pat_env_var(), clear=True):
+
+            with self.assertLogs() as context_manager:
+
+                factory = ApiClientFactory(api_secrets_filename="INVALID_SECRETS")
+                api = factory.build(InstrumentsApi)
+
+                self.assertEqual(
```

Comment:
should this assert on the good access token?

Reply:
ok i think i misunderstood the test, is this when the secrets doesn't exist rather than a bad value in it?